File: Tree/q173_ binary_search_tree_iterator.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BSTIterator:
    def __init__(self, root: TreeNode):
        self.stack = []
        self._leftmost_inorder(root)

    def _leftmost_inorder(self, root):
        while root:
            self.stack.append(root)
            root = root.left

    def next(self) -> int:
        topNode = self.stack.pop()  # 下一个最小结点
        if topNode.right:
            self._leftmost_inorder(topNode.right)
        return topNode.val

    def hasNext(self) -> bool:
        return bool(len(self.stack))


# (2)写法2：与法1相同。
class BSTIterator2:
    def __init__(self, root):
        self.stack = []
        while root:
            self.stack.append(root)
            root = root.left

    def next(self) -> int:
        topNode = self.stack.pop()
        node = topNode.right
        while node:
            self.stack.append(node)
            node = node.left
        return topNode.val

    def hasNext(self):
        return len(self.stack) > 0

File: DataStructure/Tree/q173_binary_search_tree_iterator.java
package DataStructure.Tree;

import java.util.ArrayDeque;
import java.util.Deque;

public class q173_binary_search_tree_iterator {
    Deque<TreeNode> stack = new ArrayDeque<>();

    public q173_binary_search_tree_iterator(TreeNode root) {
        inorderLeftMostTraverse(root); // 【1. 左】
    }

    // 将左孩子依次压栈 【中序 - 左根右】
    private void inorderLeftMostTraverse(TreeNode root) {
        while (root != null) {
            stack.push(root);
            root = root.left;
        }
    }

    /** @return the next smallest number */
    public int next() {
        TreeNode p = stack.pop(); // 【2. 根】
        if (p.right != null)
            inorderLeftMostTraverse(p.right); // 【3. 右 - 继续中序遍历，压栈】
        return p.val;
    }

    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        return !stack.isEmpty();
    }
}
